### PR TITLE
update flask sqlalchemy

### PR DESCRIPTION
**Pull request by utnapischtim**
- setup: increase flask-sqlalchemy version
- fix: tests LocalProxy
- change: remove click 3 compatibility
- change: add proxy file

NOTE: the [bugfix](https://github.com/inveniosoftware/invenio-db/pull/158/commits/c8024093606d9d37c19177c23aeaeb5a6f70ab49) has to be discussed before merging!

if we use the change then following two lines have to be updated:
[invenio_oauth2server.utils](https://github.com/inveniosoftware/invenio-oauth2server/blob/a66db442f7c2744f3b9f604523c5d67783231276/invenio_oauth2server/utils.py#L36)
[invenio_oauthclient.utils](https://github.com/inveniosoftware/invenio-oauthclient/blob/4a7a2d08b6d14ebf9509ad896375f709730fe0f6/invenio_oauthclient/utils.py#L307)

### CODE REVIEW

File: invenio_db/utils.py
Comment (ntarocco) [inline invenio_db/utils.py:19]:
```suggestion
def rebuild_encrypted_properties(old_key, model, properties, db=_db):
```

File: invenio_db/utils.py
Comment (ntarocco) [inline invenio_db/utils.py:16]:
```suggestion
from .shared import db as _db
```

File: invenio_db/utils.py
Comment (ntarocco) [inline invenio_db/utils.py:19]: What error was it giving? From your explanations, using the `.shared.db` (the now default value) will always fail, otherwise there was no need to change `invenio-oauthclient/server`. If that's the case, I would remove the default value, and release a major version because of the breaking change (in any case we should probably release major given the new required min version of Python and Flask).
See [changes](https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/changes/).

Was the error maybe related to any of the following documented changes?

> - The session is scoped to the current app context instead of the thread. This requires that an app context is active. This ensures that the session is cleaned up after every request.
> - An active Flask application context is always required to access session and engine, regardless of if an application was passed to the constructor. [#508](https://github.com/pallets-eco/flask-sqlalchemy/issues/508), [#944](https://github.com/pallets-eco/flask-sqlalchemy/issues/944)

**Review by ntarocco — COMMENTED** [review]
Thanks a lot for this! We should carefully test an InvenioRDM app and upgrades to make sure that any of the [changes](https://flask-sqlalchemy.palletsprojects.com/en/3.0.x/changes/) are not affecting an Invenio app.

File: invenio_db/utils.py
Comment (ntarocco) [reply invenio_db/utils.py:19]: After testing, the error is `KeyError: <weakref at 0x10cb54c70; to 'Flask' at 0x10cb5ac70>`.
It looks like that the app is not init in the tests, because of the [custom class SQLAlchemy](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/shared.py#L31).
If you change the [fixture](https://github.com/inveniosoftware/invenio-db/blob/master/tests/conftest.py#L26) to:
```diff
db = invenio_db.db = shared.db = shared.SQLAlchemy(
+      app,
        metadata=shared.MetaData(naming_convention=shared.NAMING_CONVENTION)
    )
```
then is solves **partially** the error.
I think that we need to dig more in the error.

I have also saw this in the changes:
`SQLALCHEMY_TRACK_MODIFICATIONS is disabled by default. [#727](https://github.com/pallets-eco/flask-sqlalchemy/pull/727)`
[It looks like](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/ext.py#L79) that we want it `True` instead?

Comment (ntarocco) [reply invenio_db/utils.py:19]: It might be that what is happening is that when we exit from the `app.app_context`, the internal weak reference is garbage collected.

File: setup.cfg
Comment (utnapischtim) [reply setup.cfg:41]: Yes. I tried without it, and the error is `ModuleNotFoundError: No module named 'six'` coming from `sqlalchemy_continuum/operation.py`

File: invenio_db/utils.py
Comment (utnapischtim) [reply invenio_db/utils.py:19]: > It looks like that the app is not init in the tests, because of the [custom class SQLAlchemy](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/shared.py#L31).

I thought that the db is initialized [with](https://github.com/inveniosoftware/invenio-db/blob/9ea8eb3a6f05925d26dacf76645610bccef4f278/invenio_db/ext.py#L83) by calling it [with](https://github.com/inveniosoftware/invenio-db/blob/9ea8eb3a6f05925d26dacf76645610bccef4f278/tests/test_utils.py#L39)

> the now default value) will always fail

Do you mean because of the same name of the imported and the parameter name?
It is working on my local machine and it is also not failing for postgresql on github. I have no idea what the problems with mysql8 are!

> If you change the [fixture](https://github.com/inveniosoftware/invenio-db/blob/master/tests/conftest.py#L26) to:

how do you get the `app into the db() fixture to initialize it in the custom SQLAlchemy class?

> I think that we need to dig more in the error.

I will have a look at your pointed out documented changes. It could be that this could be the root of the problem. 

> It might be that what is happening is that when we exit from the app.app_context, the internal weak reference is garbage collected.

I tried to call `InvenioDB(app, entry_point_group=False, db=db)` within the `with app.app_context()` context. The result was the same error.

> I think that we need to dig more in the error.

I will investigate deeper into the garbage collector idea and the documented changes. Hopefully it is not necessary to change the `rebuild_encrypted_properties` function

> [It looks like](https://github.com/inveniosoftware/invenio-db/blob/master/invenio_db/ext.py#L79) that we want it True instead?

the linked line sets this variable to True? I think i am missing something.

Comment (ntarocco) [reply invenio_db/utils.py:19]: > how do you get the `app into the db() fixture to initialize it in the custom SQLAlchemy class?
Just to test, I tried with adding the `app` fixture in the `db` fixture:
```
def db(app):
```

> the linked line sets this variable to True? I think i am missing something.
it looks to me that we are doing a `setdefault`: if it is already `False`, it won't be changed. To be tested!

**Comment by utnapischtim** [comment]
@ntarocco How should we proceed with this?

File: setup.cfg
Comment (ntarocco) [inline setup.cfg:36]: Shall we try to remove the upper pin here?

File: tests/test_utils.py
Comment (ntarocco) [inline tests/test_utils.py:36]: It should not be an issue, but I cannot find where the `length` params comes from in `StringEncryptedType`. Is it necessary given that it was not there earlier?

File: invenio_db/cli.py
Comment (ntarocco) [inline invenio_db/cli.py:15]: In general, what we put in the `proxies`, is our Invenio extensions. Shall we keep this proxy inside here, private, so that there is no confusion of what is `current_db`?
When I see `current_db`, the first thought that I have is that it is `invenio_db`, and not a reference to `sqlalchemy`.

File: invenio_db/utils.py
Comment (ntarocco) [inline invenio_db/utils.py:16]: I think this was changed in the wrong way?

The previous `db`  should be the same, no change. The previous `_db`, should be the `current_db`:
```suggestion
from .proxies import current_db as _db
from .shared import db
```

**Review by ntarocco — APPROVED** [review]

**Comment by ntarocco** [comment]
> @ntarocco How should we proceed with this?
> 
> The change of the signature of the `rebuild_encrypted_properties` function is backward compatible but it may be necessary to update the two usages in `invenio-oauth2server` and `invenio-oauthclient`
It looks like that there are tests there, so we can run the tests and see if they pass. If there is a change in the future, the tests should not pass.

File: setup.cfg
Comment (utnapischtim) [reply setup.cfg:36]: Normally i would say yes, we should remove the upper pin, but i would like to ask you, how do we see as early as possible if the version 4.0.0 creates problems? the github actions on this packages will be stopped after some time. So we will see the problems on a local installation, which could block work for some persons and we have to upper pin it again. 

Otherwise if we use the upper pin, we maybe see some DeprecationWarnings and we could do the same as now and have more time to not block others.

So for me, we should first see if we have a setup where we can see problems in advance before we remove the upper pin.

File: tests/test_utils.py
Comment (utnapischtim) [reply tests/test_utils.py:36]: [The](https://github.com/inveniosoftware/invenio-db/pull/158/commits/afba0be767505b0df9aaebfb310212bf78412e4e) problem is VARCHAR on mysql. Otherwise the tests are not working. Which leads me to a question: how long do we support mysql?

File: invenio_db/cli.py
Comment (utnapischtim) [reply invenio_db/cli.py:15]: hm, good point. Maybe that was also the reason it was here in the first place. I created the proxies.py file. OK i will change that back.

File: invenio_db/utils.py
Comment (utnapischtim) [reply invenio_db/utils.py:16]: here i am a little bit confused, because the `from .shared import db as _db` was your [suggestion](https://github.com/inveniosoftware/invenio-db/pull/158#discussion_r1053289837).  The `drop_alembic_version_table` function uses `current_db`. If we change `current_db` to `_db` we would have to change that too in the `drop_alembic_version_table` function. Further, as i understand the two usages should not be the same because in `rebuild_encrypted_properties` the `db` imported from `shared.py` was used and not the `_db` from the `LocalProxy`

File: tests/test_utils.py
Comment (ntarocco) [reply tests/test_utils.py:36]: I think that we should support MySQL if it does not create great issues, otherwise we might need an ORM layer at all :) There could be organization that are required to use MySQL for example.

File: invenio_db/utils.py
Comment (ntarocco) [reply invenio_db/utils.py:16]: I am sorry if I made the wrong suggestion.
What I noticed now is that we have switched the imports, which might lead to issues.
Nothing should be changed, given that the imports will be exactly the same as before the changes. The only needed change in this file is the extra param to `rebuild_encrypted_properties`

Comment (utnapischtim) [reply invenio_db/utils.py:16]: yes, but the new parameter has to come from .shared not from .proxies. they are not the same. but i will investigate a little bit more into that problem.

**Comment by fenekku** [comment]
Coordination:

This looks like it can be merged: tests are passing and maintainer has approved. I don't know that another outside reviewer could review this well. 

If @ntarocco wants to merge, go ahead.

**Comment by ntarocco** [comment]
@fenekku thanks, we are still chatting offline on these changes. I would also test these changes before merging. Almost ready :)

**Pull request merged**
